node/bindnode/node.go

```diff
@@ -0,0 +1,1099 @@
+// schema, like a reverse of PrototypeNoSchema.
+func PrototypeOnlySchema(schemaType schema.Type) ipld.NodePrototype {
+	goType := inferGoType(schemaType)
+	return prototype(goType, schemaType)
+}
+
+// from IPLD Schema field names like "foo" to Go field names like "Foo".
+func fieldNameFromSchema(name string) string {
+	return strings.Title(name)
+}
+
+func inferGoType(typ schema.Type) reflect.Type {
+	switch typ := typ.(type) {
+	case *schema.TypeString:
+		return goTypeString
+	case *schema.TypeStruct:
+		fields := typ.Fields()
+		goFields := make([]reflect.StructField, len(fields))
+		for i, field := range fields {
+			ftyp := inferGoType(field.Type())
+			if field.IsNullable() {
+				ftyp = reflect.PtrTo(ftyp)
+			}
+			if field.IsOptional() {
+				ftyp = reflect.PtrTo(ftyp)
+			}
+			goFields[i] = reflect.StructField{
+				Name: fieldNameFromSchema(field.Name()),
+				Type: ftyp,
+			}
+		}
+		return reflect.StructOf(goFields)
+	case *schema.TypeMap:
+		ktyp := inferGoType(typ.KeyType())
+		vtyp := inferGoType(typ.ValueType())
+		if typ.ValueIsNullable() {
+			vtyp = reflect.PtrTo(vtyp)
+		}
+		return reflect.MapOf(ktyp, vtyp)
+	case *schema.TypeList:
+		etyp := inferGoType(typ.ValueType())
+		if typ.ValueIsNullable() {
+			etyp = reflect.PtrTo(etyp)
+		}
+		return reflect.SliceOf(etyp)
+	case *schema.TypeUnion:
+		type goUnion struct {
+			Index int // 0..len(typ.Members)-1
+			Value interface{}
+		}
+		return reflect.TypeOf(goUnion{})
+	}
+	panic(fmt.Sprintf("%T\n", typ))
+}
+
+// Prototype implements an ipld.NodePrototype given a Go pointer type and an
+// IPLD schema type.
+//
+// In this form, it is assumed that the Go type and IPLD schema type are
+// compatible. TODO: check upfront and panic otherwise
+func Prototype(ptrType interface{}, schemaType schema.Type) ipld.NodePrototype {
+	goPtrType := reflect.TypeOf(ptrType)
+	if goPtrType.Kind() != reflect.Ptr {
+		panic("ptrType must be a pointer")
+	}
+	return prototype(goPtrType.Elem(), schemaType)
+}
+
+func prototype(goType reflect.Type, schemaType schema.Type) ipld.NodePrototype {
+	if goType.Kind() == reflect.Invalid {
+		panic("goType must be valid")
+	}
+	if schemaType == nil {
+		panic("schemaType must not be nil")
+	}
+	return &_prototype{schemaType: schemaType, goType: goType}
+}
+
+var (
+	_ ipld.NodePrototype = (*_prototype)(nil)
+
+	_ ipld.Node        = (*_node)(nil)
+	_ schema.TypedNode = (*_node)(nil)
+
+	_ ipld.NodeBuilder   = (*_builder)(nil)
+	_ ipld.NodeAssembler = (*_assembler)(nil)
+
+	_ ipld.MapAssembler = (*_structAssembler)(nil)
+	_ ipld.MapIterator  = (*_structIterator)(nil)
+
+	_ ipld.ListAssembler = (*_listAssembler)(nil)
+	_ ipld.ListIterator  = (*_listIterator)(nil)
+)
+
+type _prototype struct {
+	schemaType schema.Type
+	goType     reflect.Type // non-pointer
+}
+
+func (w *_prototype) NewBuilder() ipld.NodeBuilder {
+	return &_builder{_assembler{
+		schemaType: w.schemaType,
+		val:        reflect.New(w.goType).Elem(),
```

Comment:
So there is basically never going to be a way to avoid this alloc in here, right?  Short of, maybe, some very fancy use of `unsafe`?

Maybe an acceptable cost for this system, just thinking outloud about how docs should talk about performance of this system.

Reply:
I mean, the Builder/Assembler themselves already alloc, as they are interfaces. And this is one alloc per built Node, so I don't think it's particularly worrysome.

I am not worrying about performance just yet. My high-level intuition is that I'll be able to allocate about the same, or perhaps a little less, than the codegen. Another question will be the actual performance, where allocs are not the only factor.